core/src/main/scala/com/metamx/tranquility/tranquilizer/Tranquilizer.scala

```diff
@@ -0,0 +1,302 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.metamx.tranquility.tranquilizer
+
+import com.metamx.common.lifecycle.LifecycleStart
+import com.metamx.common.lifecycle.LifecycleStop
+import com.metamx.common.scala.Logging
+import com.metamx.common.scala.Predef._
+import com.metamx.common.scala.concurrent.loggingThread
+import com.metamx.tranquility.beam.Beam
+import com.twitter.util.Await
+import com.twitter.util.Future
+import com.twitter.util.Promise
+import com.twitter.util.Return
+import com.twitter.util.Throw
+import scala.collection.mutable.ArrayBuffer
+import scala.collection.mutable.Buffer
+
+class Tranquilizer[MessageType] private(
+  beam: Beam[MessageType],
+  maxBatchSize: Int,
+  maxPendingBatches: Int,
+  lingerMillis: Long
+) extends Logging
+{
+  require(maxBatchSize >= 1, "batchSize >= 1")
+  require(maxPendingBatches >= 1, "maxPendingBatches >= 1")
+  require(lingerMillis >= 0, "lingerMillis >= 0")
+
+  @volatile private var started: Boolean = false
+
+  case class MessageHolder(message: MessageType, future: Promise[Unit])
+
+  private val lock             : AnyRef                = new AnyRef
+  private var buffer           : Buffer[MessageHolder] = new ArrayBuffer[MessageHolder]
+  private var bufferStartMillis: Long                  = 0L
+  private var pendingBatches   : Int                   = 0
```

Comment:
can we add a comment to explain upon updating which variables we are supposed to call lock.notifyAll ?